Add: Enemy will surrender when city becomes free

GoldJohnKing opened the pull request:
(Use English only.)

- Add: Enemy will surrender when city becomes free (@GoldJohnKing).

**When merged this pull request will:**
- If enemy amount is less than (0, 1, 2, 3), city will becomes free and remaining enemies will surrender.
- If unit is a UAV, it will be destroyed immediately.
- If unit is inside vehicle, it will dismount then surrender.
- If unit is on foot, it will surrender.
- Also added `setTriggerInterval` to decrease trigger check interval (default 0.5s now 2s), it may slightly improve performance.
- I'm new on coding, and the code seems not that gentle than I thought, so this might need to be carefully reviewed.

**_TODO_**
- [x] Need an acceptable approach to get remaining units in the city.
- [x] Add mission parameter and let server admin choose if to enable this function.
- [x] AI inside vehicle won't surrender as expected.
- [x] Prevent enemies inside "Air" and "Ship" vehicles from surrendering, as they will likely fly by instead of landing.
- [x] ~~Handle remote units on headless clients such as enemy patrols.~~ All used syntax are argument global and effect global now.
- [x] Trigger detect enemies inside the same vehicle as one enemy unit, we don't want city to be free when there's a vehicle full of enemy units.
- [x] Use new syntax `configOf`.
- [x] It's rather an ugly approach right now, we should optimize the code if possible.
- [x] Use `unitIsUAV` syntax.
- [x] Discuss behavior handling of vehicles.

**Final test:**
- [x] local
- [x] server

**Screenshots**

Vdauphin reviewed (changes requested):
I really dislike this because it make useless to talk to civilian to find last enemies

This must be optionnal like this: 
Auto free cities when number of enemies is lower than:
- 0
- 1
- 2
- 3

GoldJohnKing commented:
> I really dislike this because it make useless to talk to civilian to find last enemies

Yes, I agree with you, however AI sometimes stuck inside rock or solid buildings, and bomb UAV sometimes run out of fuel and land inside grass... I used to talk to civilian to obtain information about enemy position, however they prefer to lie in most cases, even we have a high reputation. (I've read #1021, so this might be a bug?) Besides,  It's rather difficult for players to find those enemies, this is rather desperate for players as they make all efforts but to find this mission too difficult to be achieve.

> This must be optional like this:
> Auto free cities when number of enemies is lower than:
> 
> * 0
> * 1
> * 2
> * 3

Yes, I agree, I will do that.

GoldJohnKing commented on =BTC=co@30_Hearts_and_Minds.Altis/core/fnc/city/set_clear.sqf:
```diff
@@ -38,3 +38,15 @@ if (btc_final_phase) then {
+        };
+        [_x, true] call ace_captives_setSurrendered;
+    };
+} forEach (_city getVariable ["data_units", []]);
```
My fault... All I can come up with now is:

1. Use allUnits to get all units
2. Use select or inAreaArray to get units inside city radius

This could be... a mess. Would you please suggest an approach to obtain all units assigned to current city?

Vdauphin replied:
It is exactly what I would do!

Vdauphin commented:
>  agree with you, however AI sometimes stuck inside rock 

This should not happen: https://github.com/Vdauphin/HeartsAndMinds/blob/master/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/city/findPos.sqf#L45

> or solid buildings, 

Never notice glitch, may be I play too much on vanilla

>  however they prefer to lie in most cases, even we have a high reputation. 

This should not happen if reputation is above 1200 I think.
But under it may be fake or not: https://github.com/Vdauphin/HeartsAndMinds/blob/master_daily/=BTC=co@30_Hearts_and_Minds.Altis/core/fnc/info/ask.sqf#L61

> It's rather difficult for players to find those enemies, this is rather frustrating to players as they make all efforts but to find this mission too difficult to be achieve.

This fix just make sure we get the closer enemy. Personally I feel AI really magnetize by player

GoldJohnKing commented:
Hi, I've pushed a new commit, would you please check if it solves issues about above conversations?

Vdauphin reviewed (commented):
Looks cool now

GoldJohnKing commented:
Okay, this should be okay.

GoldJohnKing commented:
I'm really confused. I tried to use the same method as `btc_fnc_side_capture_officer` has used to make vehicle crew get out of the vehicle and surrender, but it just won't work...

https://github.com/Vdauphin/HeartsAndMinds/blob/5e15500ab030956569a21d615b21ccb771c6d705/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/side/capture_officer.sqf#L128

May be we can ignore those units inside vehicles or static weapons from surrendering?

Vdauphin commented:
> make vehicle crew get out of the vehicle and surrender

This code doesn't order crew to get out

GoldJohnKing commented:
I remember the officer in `btc_fnc_side_capture_officer` side-mission would automatically disembark from its vehicle and surrender if player is nearby, which only uses `doStop _captive; [_captive, true] call ace_captives_fnc_setSurrendered;`. I wonder why it doesn't work this time...

GoldJohnKing commented:
Another problem:

I see arguments of `unassignVehicle` and `doGetOut` are local, so what about those remote units on headless client, such as enemy patrols passing the city?

GoldJohnKing commented:
Hi, I've done most of the works that I can come up with. It works now, though seems to be quite a ugly approach. It's time to review and optimize the code if possible.

> > agree with you, however AI sometimes stuck inside rock
> 
> This should not happen: https://github.com/Vdauphin/HeartsAndMinds/blob/master/%3DBTC%3Dco%4030_Hearts_and_Minds.Altis/core/fnc/city/findPos.sqf#L45
> 
> > or solid buildings,
> 
> Never notice glitch, may be I play too much on vanilla

Bad news, this happens just on vanilla with only latest stable ace and cba loaded.

![Arma3_x64 2021-01-09 16-06-42](https://user-images.githubusercontent.com/26923626/104086909-aca50500-5296-11eb-977e-cf4d4b739478.png)

![Arma3_x64 2021-01-09 16-06-48](https://user-images.githubusercontent.com/26923626/104086925-c7777980-5296-11eb-89ae-d9e03028d3ef.jpg)

![Arma3_x64 2021-01-09 16-07-01](https://user-images.githubusercontent.com/26923626/104086920-baf32100-5296-11eb-9940-69afdc1834c9.png)

Vdauphin commented:
> I remember the officer in btc_fnc_side_capture_officer side-mission would automatically disembark from its vehicle and surrender if player is nearby,

This is just an AI behavior. When AI see enemies they just getout for the fight

> which only uses doStop _captive; [_captive, true] call ace_captives_fnc_setSurrendered;. I wonder why it doesn't work this time...

There is nothing asking AI to getout of vehicle here.


> Bad news, this happens just on vanilla with only latest stable ace and cba loaded.

Please don't offtopic, make an issue with step

GoldJohnKing commented on =BTC=co@30_Hearts_and_Minds.Altis/core/fnc/city/set_clear.sqf:
```diff
@@ -21,12 +22,39 @@ Author:
 private _city = btc_city_all select _id;
 _city setVariable ["occupied", false];
 
+if !(_remainEnemyUnits isEqualTo []) then {
```
Nope, this is just to make sure if `_remainEnemyUnits` is empty to avoid execution on empty string. Settings off is handled in the trigger's condition statement, in mission.sqf. (I don't know how to quote the code itself in pr...)

```
private _p_city_free_trigger = "btc_p_city_free_trigger" call BIS_fnc_getParamValue;
btc_p_city_free_trigger_condition = if (_p_city_free_trigger isEqualTo 0) then {
    "thisList isEqualTo = []" // This handles condition if setting is off
```

Vdauphin replied:
You are right, I won't do that check
and remove the useless line:
`    private _cfgVehicles = configFile >> "CfgVehicles";`

GoldJohnKing replied:
Done.

Vdauphin reviewed (changes requested):
From my point of view, the AI should not surrender if there are inside a heli or vehicle. It doesn't really make sense

GoldJohnKing commented on =BTC=co@30_Hearts_and_Minds.Altis/core/fnc/city/set_clear.sqf:
```diff
@@ -21,12 +22,38 @@ Author:
+if !(_remainEnemyUnits isEqualTo []) then {
+    {
+        private _vehicle = vehicle _x;
+        if (getNumber (configOf _vehicle >> "isUav") isEqualTo 1) then {
```
I remember this file already use `unitIsUAV`. Have you pulled the branch before review?

> From my point of view, the AI should not surrender if there are inside a heli or vehicle. It doesn't really make sense

I agree, this need further discussion.

GoldJohnKing commented:
@Vdauphin Sorry for late reply, I'm pretty busy these days... 😅💦

We need to discuss if remaining infantries should surrender while the city has at least one vehicle (including LandVehicle, Air and Ship).

Here are some situations I can come up with:

1. Infantries should surrender while city will be free, however vehicles should continue fighting.
2. Infantries should not surrender while city will not be free, until remaining vehicles are destroyed or leave the city.
3. Only with LandVehicle inside the city, infantries should not surrender while city will not be free, until those LandVehicles are destroyed or leave the city, Ship or Air vehicle inside the city neither surrender nor prevent city from becoming free.

Which one do you prefer? If you have a better plan, please enlighten me and I'll give it a try.

GoldJohnKing commented on =BTC=co@30_Hearts_and_Minds.Altis/core/fnc/city/set_clear.sqf:
```diff
@@ -21,12 +22,38 @@ Author:
+            _x setDamage 1;
+        } else {
+            if (_vehicle isKindOf "Air" || {_vehicle isKindOf "Ship"}) then {
+                _remainEnemyUnits = _remainEnemyUnits - crew _vehicle;
```
Should be resolved by vehicle surrender behavior refactor (6b8a116).

GoldJohnKing commented on =BTC=co@30_Hearts_and_Minds.Altis/core/fnc/city/set_clear.sqf:
```diff
@@ -21,12 +22,38 @@ Author:
+                        moveOut _x;
+                        [_x, true] call ace_captives_fnc_setSurrendered;
+                    } forEach _crew;
+                    _remainEnemyUnits = _remainEnemyUnits - _crew;
```
Should be resolved by vehicle surrender behavior refactor (6b8a116).

Vdauphin commented:
> 2. Infantries should not surrender while city will not be free, until remaining vehicles are destroyed or leave the city.

This one

GoldJohnKing commented:
> > 1. Infantries should not surrender while city will not be free, until remaining vehicles are destroyed or leave the city.
> 
> This one

What about UAV? There are always UAV run out of fuel and land in the grass, making it difficult for players to find and destroy it to free the city.

Vdauphin commented:
> > > 1. Infantries should not surrender while city will not be free, until remaining vehicles are destroyed or leave the city.
> > 
> > 
> > This one
> 
> What about UAV? There are always UAV run out of fuel and land in the grass, making it difficult for players to find and destroy it to free the city.

Exclude them from the search

GoldJohnKing commented:
Changed as requested, please review.
I also added units in `StaticWeapon` to surrender list, because sometimes static weapons are spawned inside "solid"/non-enterable buildings.

Vdauphin commented:
You exclude UAV because those are not AI. It is a problem due to the Arma implementation.
For static weapon, it doesn't make sense.

> because sometimes static weapons may spawned inside "solid"/non-enterable buildings.

Your pull request is to make AI surrender when there are just a few of them. So this is out of the scope of this pull request.
Make a FIX for this issue instead.

GoldJohnKing commented:
> Your pull request is to make AI surrender when there are just a few of them. So this is out of the scope of this pull request.
> Make a FIX for this issue instead.

My pull request aims at providing a workaround, which should avoid the issue that buggy AI could prevent a city from becoming free, such as unit inside rock or stuck in building, and UAV out of fuel stay silently in the grass. There are so many circumstances which might cause the issue, so make those remaining enemy surrender is the most effective way of making player less frustrated during the process.

In a word, surrender is just a method, solving the issue is the purpose.

Vdauphin commented:
> > Your pull request is to make AI surrender when there are just a few of them. So this is out of the scope of this pull request.
> > Make a FIX for this issue instead.
> 
> My pull request aims at providing a workaround, which should avoid the issue that buggy AI could prevent a city from becoming free, such as unit inside rock or stuck in building, and UAV out of fuel stay silently in the grass. There are so many circumstances which might cause the issue, so make those remaining enemy surrender is the most effective way of making player less frustrated during the process. So this pull request includes infantry, static weapon, and UAV, while does not include other types of vehicle, because vehicle can move and is much more easier to be found. It's more like a fail-safe, instead of a complete new feature.
> 
> In a word, surrender is just a method, solving the issue is the purpose. That's why I think these should be included in the same pr.

Sorry, I do not accept a workaround for those issue. I only accept fix for issue. If those issue exist please make a ticket. I don't know if you notice but all issue here are resolved by taking the time to understand why they appear. And when there are carefully understood, a fix is provided.

GoldJohnKing commented:
Ok, I get it.

GoldJohnKing commented:
Changed as requested, please review.

Vdauphin commented on =BTC=co@30_Hearts_and_Minds.Altis/core/def/mission.sqf:
```diff
@@ -84,6 +84,12 @@ btc_p_trigger = if (("btc_p_trigger" call BIS_fnc_getParamValue) isEqualTo 1) th
+btc_p_city_free_trigger_condition = if (_p_city_free_trigger isEqualTo 0) then {
+    "thisList isEqualTo []"
+} else {
+    format ["if (count thisList > %1) exitWith {false}; private _return = true; private _veh; {_veh = vehicle _x; if !(_veh isKindOf 'Man' || {unitIsUAV _veh}) exitWith {_return = false;};} forEach thisList; _return", _p_city_free_trigger, _p_city_free_trigger]
```
```
if (count thisList > %1) exitWith {false};
thisList findIf {
    private _veh = vehicle _x;
    !(
        _veh isKindOf 'Man' ||
        {unitIsUAV _veh}
    )
} isEqualTo -1
```
This looks more readable if it is inside a file under city folder.

GoldJohnKing replied:
Okay, working in progress...

GoldJohnKing replied:
> ```
> if (count thisList > %1) exitWith {false};
> thisList findIf {
>     private _veh = vehicle _x;
>     !(
>         _veh isKindOf 'Man' ||
>         {unitIsUAV _veh}
>     )
> } isEqualTo -1
> ```

I wonder if this would have impact on performance.

In my code, once a vehicle is found, it will exit immediately:
```
{
        _veh = vehicle _x;
        if !(_veh isKindOf 'Man' || {unitIsUAV _veh}) exitWith { // Exit immediately once a vehicle is found
            _return = false;
        };
    } forEach _remainEnemyUnits;
```

This might save some CPU time, because it's no need to go through the whole array.

Vdauphin reviewed (approved):
Perfect! Good job!
